src/mgr/MgrOpRequest.h

```diff
@@ -0,0 +1,116 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef OPREQUEST_H_
+#define OPREQUEST_H_
+
+#include "common/TrackedOp.h"
+#include "common/tracer.h"
+/**
+ * The MgrOpRequest takes in a MessageRef and takes over a single reference
+ * to it, which it puts() when destroyed.
+ */
+struct MgrOpRequest : public TrackedOp {
+  friend class OpTracker;
+
+public:
+  void _dump(ceph::Formatter *f) const override;
+
+private:
+  MessageRef request; /// the logical request we are tracking
+  entity_inst_t req_src_inst;
+  uint8_t hit_flag_points;
+  uint8_t latest_flag_point;
+  const char* last_event_detail = nullptr;
+
+  static const uint8_t flag_started =            1 << 0;
+  static const uint8_t flag_queued_for_module =  1 << 1;
+  static const uint8_t flag_reached_module =     1 << 2;
+
+  MgrOpRequest(MessageRef req, OpTracker *tracker);
+
+protected:
+  void _dump_op_descriptor(std::ostream& stream) const override;
+  void _unregistered() override;
+  bool filter_out(const std::set<std::string>& filters) override;
+
+public:
+  ~MgrOpRequest() override {
+    request->put();
+  }
+
+  template<class T>
+  const T* get_req() const { return static_cast<const T*>(request); }
+
+  const MessageRef get_req() const { return request; }
+  MessageRef get_nonconst_req() { return request; }
+
+  entity_name_t get_source() {
+    if (request) {
+      return request->get_source();
+    } else {
+      return entity_name_t();
```

Comment:
```suggestion
      return {};
```

Hopefully no need to repeat the type name.